### PR TITLE
Add a TaskSink internal class.

### PR DESCRIPTION
**Pull request by mykhailo-kuchma**
Add a new TaskSink class to handle the tasks creation and enqueuing.
Fix a scenario when the task could be enqueued during destruction.

Relates-To: OAM-695

Signed-off-by: Mykhailo Kuchma <ext-mykhailo.kuchma@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h [hunk 1]
```diff
@@ -132,6 +131,10 @@ class QueryMetadataJob {
 
       auto download_job = download_job_;
 
+      std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h:134]: Is this a leftover from testing? Why should we wait for 1s here?

File: olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h [hunk 2]
```diff
@@ -66,14 +67,12 @@ class QueryMetadataJob {
       QueryItemsFunc<ItemType, QueryType> query,
       FilterItemsFunc<ItemType> filter,
       std::shared_ptr<DownloadItemsJob<ItemType, PrefetchResult>> download_job,
-      std::shared_ptr<thread::TaskScheduler> task_scheduler,
-      std::shared_ptr<client::PendingRequests> pending_requests,
-      client::CancellationContext execution_context, uint32_t priority)
+      TaskSink& task_sink, client::CancellationContext execution_context,
+      thread::Priority priority)
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h:71]: Why did you change it to `thread::Priority`? What happens if the user sets some other prio than the ones defined in the `thread::Priority` enum?

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  virtual bool Put(client::TaskContext task,
+                   thread::Priority priority = thread::NORMAL) = 0;
+};
+
+class DefaultTaskSink : public TaskSink {
+ public:
+  explicit DefaultTaskSink(
+      std::shared_ptr<thread::TaskScheduler> task_scheduler)
+      : task_scheduler_(std::move(task_scheduler)),
+        pending_requests_(std::make_shared<client::PendingRequests>()),
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/TaskSink.h:45]: Do we need copies of TaskScheduler and PendingRequests? Can the TaskSink survive a VersionedLayerClient?

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+class TaskSink {
+ public:
+  virtual bool Put(client::TaskContext task,
+                   thread::Priority priority = thread::NORMAL) = 0;
+};
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/TaskSink.h:38]: Why do we need a base class?

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  DefaultTaskSink& operator=(const DefaultTaskSink&) = delete;
+
+  ~DefaultTaskSink() {
+    closed_.store(true);
+    pending_requests_->CancelAllAndWait();
+  }
+
+  void CancelTasks() { pending_requests_->CancelAll(); }
+
+  bool Put(client::TaskContext task, thread::Priority priority) override {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/TaskSink.h:58]:
```suggestion
  bool Put(client::TaskContext task, uint32_t priority) override {
```
Why should we be able to only add three priorities?

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+ private:
+  std::shared_ptr<thread::TaskScheduler> task_scheduler_;
+  std::shared_ptr<client::PendingRequests> pending_requests_;
+  std::atomic_bool closed_;
+};
+
+}  // namespace read
+}  // namespace dataservice
+}  // namespace olp
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/TaskSink.h:96]: EOF missing

File: olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h (hunk 1 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h:134]: Yes, will remove

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h (hunk 3 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/TaskSink.h:45]: It can't

File: olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h (hunk 2 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h]: reverted

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h (hunk 4 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/TaskSink.h]: Removed

File: olp-cpp-sdk-dataservice-read/src/TaskSink.h (hunk 5 above)
Comment (mykhailo-kuchma) [reply olp-cpp-sdk-dataservice-read/src/TaskSink.h]: Reverted

**Review by andescu — APPROVED** [review]

**Review by LiubovDidkivska — APPROVED** [review]

**Review by LiubovDidkivska — APPROVED** [review]

**Review by mykolamalik — APPROVED** [review]
